resource.docker_service.task_spec.container_spec.hosts is ignored

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform (and docker Provider) Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* `docker_service`, `docker_container`

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->
Only relevant parts here:
```hcl
provider "docker" {
  host = "ssh://<user@host>"
}

resource "docker_service" "example" {
  task_spec {
    container_spec {
      hosts {
        host = "host.docker.internal"
        ip   = "host-gateway"
      }
    }
  }
}
```

### Expected Behaviour

```
$ docker inspect <container-id>
...
            "ExtraHosts": [
                "host.docker.internal:host-gateway"
            ],
...
```

### Actual Behaviour

No errors or warnings, but resulting service and container do not have any extra host definitions:
```
$ docker inspect <container-id>
...
            "ExtraHosts": null,
...
```

The following workaround works:
```hcl
resource "docker_service" "example" {
  ...
  provisioner "local-exec" {
    command = "ssh <user@host> docker service update ${self.id} --host-add host.docker.internal:host-gateway"
  }
}
```

I've quoted input for `docker_service`, but `docker_container` seems to behave exactly the same. Also, I have tried normal IPv4 also, with same results.

Provider version is 2.11.0, reproducing seems to be trivial, but let me know if you want full/minimal example. Maybe a little help will be needed here in a form of some other minimal example, so I could modify it?

Switching to 2.12.0 produced whole bunch of new errors, e.g. this is during `plan` with freshly erased infra:
```
╷
│ Error: Plugin did not respond
│ 
│   with docker_network.internal_network,
│   on traefik.tf line 83, in resource "docker_network" "internal_network":
│   83: resource "docker_network" "internal_network" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadResource call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-docker_v2.12.0 plugin:

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 298 [running]:
github.com/terraform-providers/terraform-provider-docker/internal/provider.resourceDockerServiceStateUpgradeV2(0x1e7fe80, 0xc0000b7dc0, 0xc0000af320, 0x1b03e00, 0xc00064e190, 0xc0000b8a50, 0x1b125c0, 0xc00000f610)
        github.com/terraform-providers/terraform-provider-docker/internal/provider/resource_docker_service_migrators.go:26 +0x47f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).upgradeJSONState(0xc0002847b0, 0x1e7fe80, 0xc0000b7dc0, 0x1, 0xc0000af320, 0xc000438000, 0x0, 0xc0003d3aa0, 0x100d7db)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/grpc_provider.go:424 +0xd7
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).UpgradeResourceState(0xc0002847b0, 0x1e7fe80, 0xc0000b7dc0, 0xc000489300, 0xc0000b7dc0, 0x100b665, 0x1c28460)
...
Error: The terraform-provider-docker_v2.12.0 plugin crashed!
```
And this is a bit earlier, during `plan` for existing architecture and state (with single change -- new `docker_image` and `docker_service`):
```
2021-05-24T07:59:32.351+0300 [WARN]  Provider "registry.terraform.io/kreuzwerker/docker" produced an invalid plan for docker_network.internal_network, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .ipv6: planned value cty.False for a non-computed attribute
      - .ingress: planned value cty.False for a non-computed attribute
      - .ipam_driver: planned value cty.StringVal("default") for a non-computed attribute
      - .attachable: planned value cty.False for a non-computed attribute
2021-05-24T07:59:32.352+0300 [INFO]  ReferenceTransformer: reference not found: "local.test_2"
...
2021-05-24T07:59:32.359+0300 [DEBUG] provider.terraform-provider-docker_v2.12.0: PANIC: interface conversion: interface {} is nil, not map[string]interface {}
2021-05-24T07:59:32.359+0300 [DEBUG] provider.terraform-provider-docker_v2.12.0: 
2021-05-24T07:59:32.359+0300 [DEBUG] provider.terraform-provider-docker_v2.12.0: goroutine 136 [running]:
2021-05-24T07:59:32.359+0300 [DEBUG] provider.terraform-provider-docker_v2.12.0: github.com/terraform-providers/terraform-provider-docker/internal/provider.resourceDockerServiceStateUpgradeV2(0x1e7fe80, 0xc0000489c0, 0xc000575290, 0x1b03e00, 0xc0004b2090, 0xc00035e210, 0x1b125c0, 0xc0002e81c8)
2021-05-24T07:59:32.359+0300 [DEBUG] provider.terraform-provider-docker_v2.12.0:        github.com/terraform-providers/terraform-provider-docker/internal/provider/resource_docker_service_migrators.go:26 +0x47f
```
Results above were achieved with following versions:
```
$ terraform -version
Terraform v0.15.3
on darwin_amd64
+ provider registry.terraform.io/gitlabhq/gitlab v3.6.0
+ provider registry.terraform.io/hashicorp/kubernetes v2.2.0
+ provider registry.terraform.io/hashicorp/local v2.1.0
+ provider registry.terraform.io/hashicorp/null v3.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/kreuzwerker/docker v2.12.0
```
Terraform hinted there was new minor release, so I did an upgrade and tried with latest downloadable version, with basically same results.
```
$ terraform -version
Terraform v0.15.3
# rest is the same
```
These errors do not seem like being related to the original one (at least not on the same level), probably deserve separate issue?

Hey @mavogel, here is minimal `.tf` to reproduce:
```
terraform {
  required_providers {
    docker = {
      source  = "kreuzwerker/docker"
      version = "2.11.0"
    }
  }
}

provider "docker" {
  host = "ssh://docker-swarm-server"
}

resource "docker_service" "example" {
  name = "example"
  task_spec {
    container_spec {
      image = "scratch"
    }
  }
}
```
Some additional info:
* `terraform -v`:
```
Terraform v0.15.4
on darwin_amd64
```
* `docker-swarm-server` -- just an `Ubuntu 20.04.2 LTS` instance with `Docker version 20.10.6, build 8728dd2`, I have passwordless setup in my `~/.ssh/config`, nothing fancy

And steps to reproduce:
1. put the content above into new `main.tf` file within some empty directory
2. do `terraform init`:
```
Initializing the backend...

Initializing provider plugins...
- Finding kreuzwerker/docker versions matching "2.11.0"...
- Installing kreuzwerker/docker v2.11.0...
- Installed kreuzwerker/docker v2.11.0 (self-signed, key ID 24E54F214569A8A5)
...
```
3. do `terraform apply -auto-approve`:
```
# lots of output skipped
Plan: 1 to add, 0 to change, 0 to destroy.
docker_service.example: Creating...
docker_service.example: Creation complete after 5s [id=03vb5mhwes6l8j8z5y1vi847f]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```
4. Edit version in `main.tf`:
```
version = "2.12.0"
```
5. do `terraform init -upgrade`:
```
Initializing the backend...

Initializing provider plugins...
- Finding kreuzwerker/docker versions matching "2.12.0"...
- Installing kreuzwerker/docker v2.12.0...
- Installed kreuzwerker/docker v2.12.0 (self-signed, key ID 24E54F214569A8A5)
...
```
6. run any terraform command that involves state, e.g. `terraform plan`:
```
Stack trace from the terraform-provider-docker_v2.12.0 plugin:

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 29 [running]:
github.com/terraform-providers/terraform-provider-docker/internal/provider.resourceDockerServiceStateUpgradeV2(0x1e7fe80, 0xc0002553c0, 0xc0000feab0, 0x1b03e00, 0xc0006d1c30, 0xc0000ca4d0, 0x1b125c0, 0xc00000ecc8)
        github.com/terraform-providers/terraform-provider-docker/internal/provider/resource_docker_service_migrators.go:26 +0x47f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).upgradeJSONState(0xc00000cf30, 0x1e7fe80, 0xc0002553c0, 0x1, 0xc0000feab0, 0xc00047c000, 0x0, 0xc0005b3aa0, 0x100d7db)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/grpc_provider.go:424 +0xd7
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).UpgradeResourceState(0xc00000cf30, 0x1e7fe80, 0xc0002553c0, 0xc000593020, 0xc0002553c0, 0x100b665, 0x1c28460)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/grpc_provider.go:300 +0x2dc
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).UpgradeResourceState(0xc000627e00, 0x1e7ff28, 0xc0002553c0, 0xc0000a49b0, 0xc000627e00, 0xc0000c9e60, 0xc0001d4ba0)
        github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/server/server.go:281 +0xd8
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_UpgradeResourceState_Handler(0x1c670e0, 0xc000627e00, 0x1e7ff28, 0xc0000c9e60, 0xc00086b140, 0x0, 0x1e7ff28, 0xc0000c9e60, 0xc0000f8000, 0xa77)
        github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:308 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001ffdc0, 0x1e89838, 0xc000082a80, 0xc00025c800, 0xc0001a4bd0, 0x241c7a0, 0x0, 0x0, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0001ffdc0, 0x1e89838, 0xc000082a80, 0xc00025c800, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc00009a1f0, 0xc0001ffdc0, 0x1e89838, 0xc000082a80, 0xc00025c800)
        google.golang.org/grpc@v1.32.0/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd

Error: The terraform-provider-docker_v2.12.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
╷
│ Error: Plugin did not respond
│ 
│   with docker_service.example,
│   on main.tf line 14, in resource "docker_service" "example":
│   14: resource "docker_service" "example" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).UpgradeResourceState call. The plugin logs may contain more
│ details.
╵
```

--- a/internal/provider/resource_docker_container_funcs.go
+++ b/internal/provider/resource_docker_container_funcs.go
@@ -97,7 +97,7 @@ func resourceDockerContainerCreate(ctx context.Context, d *schema.ResourceData, 
 	}
 	extraHosts := []string{}
 	if v, ok := d.GetOk("host"); ok {
-		extraHosts = extraHostsSetToDockerExtraHosts(v.(*schema.Set))
+		extraHosts = extraHostsSetToContainerExtraHosts(v.(*schema.Set))
 	}
 
 	extraUlimits := []*units.Ulimit{}

--- a/internal/provider/resource_docker_container_structures.go
+++ b/internal/provider/resource_docker_container_structures.go
@@ -169,7 +169,7 @@ func ulimitsToDockerUlimits(extraUlimits *schema.Set) []*units.Ulimit {
 	return retExtraUlimits
 }
 
-func extraHostsSetToDockerExtraHosts(extraHosts *schema.Set) []string {
+func extraHostsSetToContainerExtraHosts(extraHosts *schema.Set) []string {
 	retExtraHosts := []string{}
 
 	for _, hostInt := range extraHosts.List() {

--- a/internal/provider/resource_docker_service_structures.go
+++ b/internal/provider/resource_docker_service_structures.go
@@ -288,7 +288,8 @@ func flattenServiceHosts(in []string) *schema.Set {
 	out := make([]interface{}, len(in))
 	for i, v := range in {
 		m := make(map[string]interface{})
-		split := strings.Split(v, ":")
+		split := strings.Split(v, " ")
+		log.Println("[DEBUG] got service hostnames to split:", split)
 		m["host"] = split[0]
 		m["ip"] = split[1]
 		out[i] = m
@@ -856,7 +857,7 @@ func createContainerSpec(v interface{}) (*swarm.ContainerSpec, error) {
 				}
 			}
 			if value, ok := rawContainerSpec["hosts"]; ok {
-				containerSpec.Hosts = extraHostsSetToDockerExtraHosts(value.(*schema.Set))
+				containerSpec.Hosts = extraHostsSetToServiceExtraHosts(value.(*schema.Set))
 			}
 			if value, ok := rawContainerSpec["dns_config"]; ok {
 				containerSpec.DNSConfig = &swarm.DNSConfig{}
@@ -1334,4 +1335,19 @@ func mapSetToPlacementPlatforms(stringSet *schema.Set) []swarm.Platform {
 	}
 
 	return ret
+}
+
+func extraHostsSetToServiceExtraHosts(extraHosts *schema.Set) []string {
+	retExtraHosts := []string{}
+
+	for _, hostInt := range extraHosts.List() {
+		host := hostInt.(map[string]interface{})
+		ip := host["ip"].(string)
+		hostname := host["host"].(string)
+		// the delimiter is a 'space' + hostname and ip are switched
+		// see https://github.com/kreuzwerker/terraform-provider-docker/issues/202#issuecomment-847715879
+		retExtraHosts = append(retExtraHosts, ip+" "+hostname)
+	}
+
+	return retExtraHosts
 }

--- a/internal/provider/resource_docker_service_structures.go
+++ b/internal/provider/resource_docker_service_structures.go
@@ -290,8 +290,8 @@ func flattenServiceHosts(in []string) *schema.Set {
 		m := make(map[string]interface{})
 		split := strings.Split(v, " ")
 		log.Println("[DEBUG] got service hostnames to split:", split)
-		m["host"] = split[0]
-		m["ip"] = split[1]
+		m["ip"] = split[0]
+		m["host"] = split[1]
 		out[i] = m
 	}
 	taskSpecResource := resourceDockerService().Schema["task_spec"].Elem.(*schema.Resource)

